src/python/pants/backend/helm/target_types.py

```diff
@@ -385,6 +388,39 @@ def all_helm_artifact_targets(all_targets: AllTargets) -> AllHelmArtifactTargets
 # -----------------------------------------------------------------------------------------------
 
 
+class HelmDeploymentChartField(StringField, AsyncFieldMixin):
+    alias = "chart"
+    # TODO Will be made required in next release
+    required = False
+    help = help_text(
+        f"""
+        The address of the `{HelmChartTarget.alias}` or `{HelmArtifactTarget.alias}`
+        that will be used for this deployment.
+        """
+    )
+
+    def to_address_input(self) -> AddressInput | None:
+        if self.value:
+            return AddressInput.parse(
+                self.value,
+                relative_to=self.address.spec_path,
+                description_of_origin=f"the `{self.alias}` field in the `{HelmDeploymentTarget.alias}` target {self.address}",
+            )
+
+        warn_or_error(
```

Comment:
On first reading, I don't understand why this warning is here? Why is this a deprecated codepath when this the new field? This requires at least a comment, or ideally moving this warning to a more obvious location?

Reply:
Yeah, not the most obvious place to put it. The method is called in two different places: When obtaining the chart instance (with its sources) for a given deployment and when inferring the dependencies of the helm deployment (so the chart address is listed among those).

It relies one the fact that I would like to make the field required in the subsequent release so this method would never return `None`. So now the most common case for reaching this point in the code would be that users have upgraded to this release and therefore users are not using this field yet.

Being called from two different execution paths, thought that having it here would help with memoizing the call to `warn_or_error` and the message would only be displayed one time.